llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -3297,6 +3297,35 @@ EVT RISCVTargetLowering::getOptimalMemOpType(
   return TargetLowering::getOptimalMemOpType(Op, FuncAttributes);
 }
 
+template <typename Instr>
+static bool
+needsExplicitAddressingModeAtomics(const Instr *I,
+                                   const RISCVSubtarget &Subtarget) {
+  const DataLayout &DL = I->getModule()->getDataLayout();
+  bool IsCheriPurecap = RISCVABI::isCheriPureCapABI(Subtarget.getTargetABI());
+  // The atomic RMW instructions are only available in capmode, so for the
+  // hybrid ABI we have expand these operations to use LR/SC (which are
+  // available even when not in capmode). However, the TLI hooks for
+  // AtomicExpansionKind::LLSC are not implemented so we use CmpXChg.
+  return DL.isFatPointer(I->getPointerOperand()->getType()) != IsCheriPurecap;
+}
+
+bool RISCVTargetLowering::shouldInsertFencesForAtomic(
+    const Instruction *I) const {
+  // The CHERI instructions with explicit addressing mode are always relaxed
+  // so we need to insert fences when using them. We use them for capability
+  // pointers in non-capability mode and integer pointers in capmode.
+  // Since we don't know the value of capmode here, we approximate it by
+  // comparing the address space ot the pointer operand to the default
+  // globals address space of the module data layout.
```

Comment:
RISCVTargetLowering has a Subtarget in it, which has the ABI.

Reply:
Ah yeah that's a stale comment from before when I realized that I can access the subtarget.

Reply:
Oh yeah you actually use it below :)